compiler/lib/effects.ml

```diff
@@ -482,67 +557,72 @@ let cps_last ~st ~alloc_jump_closures pc (last : last) ~k : instr list * last =
             @ (Let (exn_handler, Prim (Extern "caml_pop_trap", [])) :: body)
           , branch ))
 
-let cps_instr ~st (instr : instr) : instr =
-  match instr with
-  | Let (x, Closure (params, (pc, _))) when Var.Set.mem x st.cps_needed ->
+let rewrite_instr ~target ~st (instr : instr) : instr list =
+  match target, instr with
+  | (`Cps | `Direct_style), Let (x, Closure (_, (pc, _))) when Var.Set.mem x st.cps_needed
+    ->
+      (* When CPS-transforming with double translation enabled, there are no closures in
+         code that requires transforming, due to lambda lifiting. *)
+      assert (not (double_translate ()));
       (* Add the continuation parameter, and change the initial block if
          needed *)
-      let k, cont = Hashtbl.find st.closure_info pc in
+      let cps_params, cps_cont = Hashtbl.find st.closure_info pc in
       st.in_cps := Var.Set.add x !(st.in_cps);
-      Let (x, Closure (params @ [ k ], cont))
-  | Let (x, Prim (Extern "caml_alloc_dummy_function", [ size; arity ])) -> (
+      [ Let (x, Closure (cps_params, cps_cont)) ]
+  | ( (`Cps | `Direct_style)
+    , Let (x, Prim (Extern "caml_alloc_dummy_function", [ size; arity ])) ) -> (
       match arity with
       | Pc (Int a) ->
-          Let
-            ( x
-            , Prim
-                (Extern "caml_alloc_dummy_function", [ size; Pc (Int (Targetint.succ a)) ])
-            )
+          [ Let
+              ( x
+              , Prim
+                  ( Extern "caml_alloc_dummy_function"
+                  , [ size; Pc (Int (Targetint.succ a)) ] ) )
+          ]
       | _ -> assert false)
-  | Let (x, Apply { f; args; _ }) when not (Var.Set.mem x st.cps_needed) ->
+  | (`Cps | `Direct_style), Let (x, Apply { f; args; _ })
+    when not (Var.Set.mem x st.cps_needed) ->
       (* At the moment, we turn into CPS any function not called with
          the right number of parameter *)
-      assert (Global_flow.exact_call st.flow_info f (List.length args));
-      Let (x, Apply { f; args; exact = true })
-  | Let (_, (Apply _ | Prim (Extern ("%resume" | "%perform" | "%reperform"), _))) ->
+      assert (
+        (* If this function is unknown to the global flow analysis, then it was
+           introduced by the lambda lifting and we don't have exactness info any more. *)
+        Var.idx f >= Var.Tbl.length st.flow_info.info_approximation
+        || Global_flow.exact_call st.flow_info f (List.length args));
+      [ Let (x, Apply { f; args; exact = true }) ]
+  | `Cps, Let (x, Prim (Extern "caml_assume_no_perform", [ Pv f ])) ->
```

Comment:
This case is not supposed to happen in direct style. So maybe we can just stop matching on the `target`?
Or we could move it to a separate function:
```ocaml
let cps_instr ~st instr =
  match instr with
  | Let (x, Prim (Extern "caml_assume_no_perform", [ Pv f ])) ->
      ...
  | _ ->
      [rewrite_instr ~st instr]
```